django/tasks/base.py

```diff
@@ -0,0 +1,247 @@
+        if result.task.func != self.func:
+            raise TaskIntegrityError(
+                f"Task does not match (received {result.task.module_path!r})"
+            )
+        return result
+
+    async def aget_result(self, result_id):
+        """See get_result()."""
+        result = await self.get_backend().aget_result(result_id)
+        if result.task.func != self.func:
+            raise TaskIntegrityError(
+                f"Task does not match (received {result.task.module_path!r})"
+            )
+        return result
+
+    def call(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return async_to_sync(self.func)(*args, **kwargs)
+        return self.func(*args, **kwargs)
+
+    async def acall(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return await self.func(*args, **kwargs)
+        return await sync_to_async(self.func)(*args, **kwargs)
+
+    def get_backend(self):
+        from . import tasks
+
+        return tasks[self.backend]
+
+    @property
+    def module_path(self):
+        return f"{self.func.__module__}.{self.func.__qualname__}"
+
+
+def task(
+    function=None,
+    *,
+    priority=DEFAULT_PRIORITY,
+    queue_name=DEFAULT_QUEUE_NAME,
+    backend=DEFAULT_TASK_BACKEND_ALIAS,
+    enqueue_on_commit=None,
+    takes_context=False,
+):
+    from . import tasks
+
+    def wrapper(f):
+        return tasks[backend].task_class(
+            priority=priority,
+            func=f,
+            queue_name=queue_name,
+            backend=backend,
+            enqueue_on_commit=enqueue_on_commit,
+            takes_context=takes_context,
+            run_after=None,
+        )
+
+    if function:
+        return wrapper(function)
+    return wrapper
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class TaskError:
+    exception_class_path: str
+    traceback: str
+
+    @property
+    def exception_class(self):
+        # Lazy resolve the exception class.
+        exception_class = import_string(self.exception_class_path)
+
+        if not isclass(exception_class) or not issubclass(
+            exception_class, BaseException
+        ):
+            raise ValueError(
+                f"{self.exception_class_path!r} does not reference a valid exception."
+            )
+        return exception_class
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class TaskResult:
+    task: Task
+
+    id: str  # Unique identifier for the task result.
+    status: ResultStatus
+    enqueued_at: Optional[datetime]  # Time the task was enqueued.
+    started_at: Optional[datetime]  # Time the task was started.
+    finished_at: Optional[datetime]  # Time the task was finished.
+
+    # Time the task was last attempted to be run.
+    last_attempted_at: Optional[datetime]
+
+    args: list  # Arguments to pass to the task function.
+    kwargs: Dict[str, Any]  # Keyword arguments to pass to the task function.
```

Comment:
Instead of expecting users to do
```python
TaskResult(
    ...
    args=json_normalize(args),
    kwargs=json_normalize(kwargs),
    ...
)
```
I think it might be a good idea for us to apply `json_normalize` to `args` `kwargs` when this is initialized
We currently document this "round-trip" json normalization and we expect other backends to do this, so it would be easier for us to do it for them

If not, we should probably document `json_normalize` and document the expectation to use it

Note that `args` and `kwargs` are missing documentation in `TaskResult` currently

Reply:
It's only backend developers who would need to do this, where the backend doesn't serialize itself down to JSON for storage (where this would already be done). Always doing it in the constructor would be a lot of unnecessary extra serialization.

Reply:
Not sure I am following how this would be extra serialization rather than moving the same serialization to a different place
The main thing that concerns me is we are documenting that this would always be happening, but whether this happens or not is currently based on the implementation of the backend's `enqueue()`.